### DISCUSSION TITLE
New Follower/Following List

### DISCUSSION THREAD

**Discussion by MariaJsn** [sbertix/Swiftagram#153] [Q&A] [answered]
Hi Sbertix,

Is there any way to get if there is any change in followers/following list since last fetch.

When I check the discussions I saw below item. But I could not understand how I can use this logic.
Please can you help me?

https://github.com/sbertix/Swiftagram/discussions/118 

Thanks,
Maria

**Comment by sbertix** [comment, marked as answer]
Hey @mariajsn,

All you have to do is just cache the followers you get in the first request (using `Realm`, `CoreData`, `UserDefaults`, or whatever you're relying on) and compare them with the ones you get on refresh. 

If you only care to know whether a change occurred or not, simply pass the same rank token in every refresh: only fetch 1 page, if you get a different `next_id` it means the list has changed.

**Reply by MariaJsn** [reply]
Thanks for your answer Sbertix.
So you mean if I pass next_id (which is getting from first request) when requesting (later than first request) follower/following list, It will be return the same next_id if there are not any changes. If there is a change in list it will return different next_id.

Am I correct?

And is there any way to get only changed items?

Thanks,
Maria

**Reply by sbertix** [reply]
Not really @MariaJsn 

The first time you'll have to do something like… 

```swift
let secret: Secret = /* some `Secret` */
// Assuming this is some sort of constant between
// launches, e.g. a string literal. 
let token = "some.constant.rank.token"
// Exhaust all your followers.
Endpoint.Friendship.following(secret.identifier, rank: token)
   .unlocking(with: secret)
   .task(
      maxLength: .max,
      onChange: {  output in
         // Do what you need to do with the actual output
         // then store the first (ONLY THE FIRST)
         // next max id.
         let nextIdentifier = output.pagination.next
      }
   )
   .resume()
```

Next time you need to check for changes just do…

```swift
// The old next identifier for the first page.
let firstPageNextIdentifier: String = /* some `String` */
// Fetch one page only.
Endpoint.Friendship.following(secret.identifier, rank: token)
   .unlocking(with: secret)
   .task(
      maxLength: 1,
      onChange: {  output in
         // If the next identifier is equal to the one you've cached
         // than you can assume no changes were made.
         guard output.pagination.next != firstPageNextIdentifier else { return }
         // Fetch all followers again. 
         // There's no way to actually return only updated ones.
      }
   )
   .resume()
```

This **only** works if `token` remains the same and it is passed as the `rank` argument.

**Reply by bezpalyi** [reply]
@sbertix 
I tested the above way, but `output.pagination.next` is always nil.

```let token = "TOKEN_FOR_FOLLOWINGS"
        Endpoint.Friendship.followed(by: secret.identifier, rank: token)
            .unlocking(with: secret)
            .task(maxLength: 1) { (result) in
                switch result {
                case .success(let userCollection):
                    if let error = userCollection.error {
                        ///
                    } else {
                        userCollection.pagination.next
                    }
                case .failure(let error):
                    comp(.failure(error))
                }
            }
            .resume()

Could you let me know why `pagination.next` is nil?
Btw @MariaJsn

**Reply by sbertix** [reply]
Hey @bezpalyi, 

I'm assuming you're expecting more followers than the one returned, meaning `userCollection.pagination.next` is not `nil` cause you reached the end, correct? 
What version of **Swiftagram** are you on cause I feel like this is something I've already fixed, but I might be mistaken. 
Either way, as the entire pagination mechanism gets a rework, this is not gonna happen in `5.0.0`. 
For now you should consider accessing the next pagination parameter directly from the `userCollection.wrapper()`.

**Reply by bezpalyi** [reply]
@sbertix thanks for your kind reply.
I was going to use [v5.0.0-beta.3](https://github.com/sbertix/Swiftagram/releases/tag/5.0.0-beta.3) but there was some changes of usage.
So that I tried to reference your example. 
But I see following errors.
Could you update your example project?
![Screen Shot 2021-04-07 at 5 14 50 AM](https://user-images.githubusercontent.com/81933764/113779935-05cee880-9761-11eb-891d-1e9f51831610.png)

**Reply by sbertix** [reply]
Hey @bezpalyi,

The Example is related to beta 3, so you can just select the commit of the dependency and update the Example project locally. 
Please keep in mind no issues and discussions will be entertained about beta versions, especially this one, as almost everything is still subject to change (previous beta versions just added compatibility with the new **ComposableRequest**). 

I was also able to check the pagination next identifier works fine in the current **release** version of **Swiftagram**, so you have no need for the **development** branch, right now. 
Mostly likely what's happening is that your account is getting rate limited, cause you're fetching too many requests.

And please, no more off-topic replies to this discussion @bezpalyi.
